finalfusion-ffi/src/lib.rs

```diff
@@ -0,0 +1,67 @@
+use std::cell::RefCell;
+use std::ffi::CStr;
+use std::ffi::CString;
+use std::ffi::OsStr;
+use std::fs::File;
+use std::io::BufReader;
+use std::os::raw::c_char;
+use std::os::unix::ffi::OsStrExt;
+use std::ptr;
+
+use finalfusion::prelude::*;
+
+thread_local! {
+    static ERROR: RefCell<CString> = RefCell::new(CString::new(Vec::new()).unwrap());
+}
+
+fn update_error(err: &str) {
+    ERROR
+        .with(|e| e.replace(CString::new(err.as_bytes()).expect("Error string contains nul byte")));
+}
+
+/// Return the last error message.
+///
+/// The returned pointer is owned by the library and is valid until
+/// the next function call that can result in an error.
+#[no_mangle]
+pub extern "C" fn ff_error() -> *const c_char {
+    ERROR.with(|f| f.borrow().as_ptr())
+}
+
+#[no_mangle]
+pub extern "C" fn ff_read_embeddings(
+    filename: *const c_char,
+) -> *mut Embeddings<VocabWrap, StorageWrap> {
```

Comment:
I would prefer to support all formats like we do with finalfusion-python by having an enum.
https://github.com/finalfusion/finalfusion-ffi/blob/9e49af45e2c3a63858c8db0ab8fe378728016ef8/libfinalfusion/src/embeddings.rs#L22-L52 https://github.com/finalfusion/finalfusion-ffi/blob/9e49af45e2c3a63858c8db0ab8fe378728016ef8/libfinalfusion/src/embeddings_ffi.rs#L18-L27

Reply:
We should have different functions for different types of embeddings.

Reply:
I'm fine with that. We'd probably anyways want the `View` / `NonView` distinction if we decide to expose e.g. similarity and analogy queries.